### PR TITLE
[dim][regression] Explicit interface override

### LINKED ISSUE #17515: Explicit interface bug introduced in Mono 6.4
For classes with multiple interfaces, such as:

    public interface IName
    {
      string Name { get; }
    }
    public interface IOther<T>
    {
      T other { get; }
      string Name { get; }
    }
    public class Name1 : IName
    {
      public string Name { get { return "ClassName"; } }
      string IName.Name { get { return "InterfaceName"; } }
    }
    public class Name2 : IName, IOther<int>
    {
      public string Name { get { return "ClassName"; } }
      string IName.Name { get { return "InterfaceName"; } }
      public int other { get { return 43; } }
    }

Explicit method resolution does not work since Mono 6.4
Output of this code:

    var name1 = new Name1();
    var name2 = new Name2();
    IName iName1 = name1;
    IName iName2 = name2;
    Console.WriteLine(iName1.Name);
    Console.WriteLine(iName2.Name);

Is 
InterfaceName
ClassName

instead of
InterfaceName
InterfaceName

### PR DESCRIPTION
**Pull request by thaystg**
Precedence order:
-> Override interface method in class
-> Method in class
-> DIM

Fixes #17515

### CODE REVIEW

File: mono/tests/override-interface.il [hunk 1]
```diff
@@ -0,0 +1,722 @@
+
```
Comment (marek-safar) [inline mono/tests/override-interface.il]: Does it need to be .il test?

Comment (thaystg) [reply mono/tests/override-interface.il]: Moved to cs.

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins backport 2019-10

**Comment by marek-safar** [comment]
@monojenkins backport 2019-08

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by madewokherd** [comment]
The test case introduced here does not work in Microsoft .NET on Windows.

```
Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.TypeLoadException: Non-abstract, non-.cctor method in an interface.
   at Test.test_0_dim_override()
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at TestDriver.RunTests(Type type, String[] args, TestDriverReporter reporter)
   at TestDriver.RunTests(Type type, String[] args)
   at Test.Main(String[] args)
```

**Comment by lambdageek** [comment]
@madewokherd default interface methods depend on features that are in .NET Standard 2.1 (which is, effectively... recent Mono and .NET Core 3.0, but not .NET Framework) see https://devblogs.microsoft.com/dotnet/building-c-8-0/ "Platform Dependencies" at the bottom of the post.

**Pull request merged**